{% extends "layout.html" %}
{% block body %}
<script src="{{ url_for("static", filename="upload.js") }}"></script>
<h1>Upload</h1>
<h2>Select a triPOD formatted input file</h2>
<hr>
<form id="options" enctype="multipart/form-data" action="" method=post> 
<table class="table">

<tr class="rowcommon"> 
<td class="cell2 cell2width">Select a file to upload</td> 
<td class="cell2"><input type="file" name="file" size="25">
</tr> 

<tr class="rowcommon"> 
<td class="cell2 cell2width">Statistical significance threshold for p-value</td> 
<td class="cell2"><input name="alpha"  type="text" size="10" value="0.1"></td> 
</tr> 

<tr class="rowcommon"> 
<td class="cell2 cell2width">Reference genome build</td> 
<td class="cell2"> 
<select name="build"> 
<option value="hg16_centromeres.txt">hg16</option> 
<option value="hg17_centromeres.txt">hg17</option> 
<option value="hg18_centromeres.txt" selected>hg18</option> 
<option value="hg19_centromeres.txt">hg19(b37)</option> 
</select> 
</td> 
</tr> 

<tr class="rowcommon"> 
<td class="cell2 cell2width">Gender of sample</td> 
<td class="cell2"> 
<select name="gender"> 
<option value="M">Male</option> 
<option value="F">Female</option> 
<option value="NA" selected>NA</option> 
</select> 
</td> 
</tr> 

<tr class="rowcommon"> 
<td class="cell2 cell2width">Submit query. Note that large files will take some time to upload.</td> 
<td class="cell2"><input id="submit" type=submit value=Upload></td> 
</tr>  

<div id="submitted"><p><a>File upload is in progress. Please wait.</a></div>

<tr class="rowcommon">
<td class="cell2 cell2width">Changing the detection methods used will alter your results.</td>
<td class="cell2">
<a href="#" id="show_hide">Show/Hide detection methods</a>

</td>

</table>

<div class="extras">
<table class="table"> 

<tr class="rowcommon"> 
<td class="cell1 cell1width">NoCall rate</td> 
<td class="cell2 cell2width">Samples must have a NoCall rate below this threshold</td> 
<td class="cell2"><input name="nocall"  type="text" size="10" value="0.3"></td> 
</tr> 

<tr class="rowcommon"> 
<td class="cell1 cell1width">POD</td> 
<td class="cell2 cell2width">Parent of origin detection method</td> 
<td class="cell2"> 
<select name="pod"> 
<option value="pod" selected>Yes</option> 
<option value="nopod">No</option> 
</select> 
</td> 
</tr> 

<tr class="rowcommon"> 
<td class="cell1 cell1width">PODhd</td> 
<td class="cell2 cell2width">Homozygous deletion detection method</td> 
<td class="cell2"> 
<select name="podhd"> 
<option value="hd">Yes</option> 
<option value="nohd" selected>No</option> 
</select> 
</td> 
</tr> 

<tr class="rowcommon"> 
<td class="cell1 cell1width">PODmi1</td> 
<td class="cell2 cell2width">Single Mendelian error detection method</td> 
<td class="cell2"> 
<select name="podmi1"> 
<option value="mi1">Yes</option> 
<option value="nomi1" selected>No</option> 
</select> 
</td> 
</tr> 

<tr class="rowcommon"> 
<td class="cell1 cell1width">PODcr</td> 
<td class="cell2 cell2width">Outlier B allele frequency detection method</td> 
<td class="cell2"> 
<select name="podcr"> 
<option value="podcr">Yes</option> 
<option value="nopodcr" selected>No</option> 
</select> 
</td> 
</tr> 

</table> 
</div>

</form> 
<hr>
  <section id="description">
  <a class="subtle">This software program detects chromosomal abnormalities in parent-child trio-based microarray data. It has been shown to provide improved sensitivity for abnormality detection, with substantial improvement in detection of low-level mosaic abnormalities, as described in </a><a class="subtle" href="http://www.biomedcentral.com/1471-2164/14/367/">Joseph D. Baugher, Benjamin D. Baugher, Matthew D. Shirley, and Jonathan Pevsner. Sensitive and specific detection of mosaic chromosomal abnormalities using the Parent-of-Origin-based Detection (POD) method. BMC Genomics 2013.</a>
  </section>
  <p>
    <hr>
    <h2>Input file formatting</h2>
    <hr>
  <section id="formatting">
    <a>The input file must be tab delimited, sorted by chromosome and position, and in the following order (columns):</a>
    <pre>SNP_Name  Chromosome  Position  Father.GType  Father.BAF  Father.LRR  
  Mother.GType  Mother.BAF  Mother.LRR  Child.GType Child.BAF Child.LRR</pre>
    <a>A header line is expected and is used to extract sample names (i.e. Sample1.GType = Sample1), but not to determine column identity. Download sample data </a><a href="{{ url_for('static',filename='sampledata.txt') }}">here.</a>
    <hr>
    <h3>Genotype formatting</h3>
    <hr>
    <a>The genotypes must be AA, AB, BB, NC (or NoCall).
triPOD has been developed using genotypes annotated by the Illumina method described in "TOP/BOT" Strand and "A/B" Allele. If converting from HapMap format (ATCG) to Illumina format (AB), use simple replacement as follows:</a>
    <pre>AA = AA, TT; BB = CC, GG; AB = AC, AG, TC, TG; -- = NC. </pre>
  </section>
{% endblock %}
